text/3735-field-projections.md

```diff
@@ -0,0 +1,1541 @@
+// not have a field `inner`.
+type X = field_of!(Container<Bar>, inner);
+
+struct Baz {
+    inner: Foo,
+}
+
+// this refers to the field `inner` of `Baz`.
+type Y = field_of!(Container<Baz>, inner);
+// it has the following implementation of `Field`:
+impl Field for Y {
+    type Base = Container<Baz>;
+    type Type = Container<Foo>;
+
+    const OFFSET: usize = offset_of!(Baz, inner);
+}
+```
+
+#### `Field` Trait
+
+The field trait is added to `core::marker` and cannot be implemented manually.
+
+```rust
+pub trait Field {
+    type Base: ?Sized;
+    type Type: ?Sized;
+
+    const OFFSET: usize;
+}
+```
+
+The compiler automatically implements it for all [field types]. Users of the trait are allowed to rely
+on the associated types and constants in `unsafe` code. So for example this piece of code is sound:
+
+```rust
+fn get_field<F: Field>(base: &F::Base) -> &F::Type
+where
+    // required to be able to `cast`
+    F::Type: Sized,
+    F::Base: Sized,
+{
+    let ptr: *const Base = base;
+    let ptr: *const u8 = base.cast::<u8>();
+    // SAFETY: `ptr` is derived from a reference and the `Field` trait is guaranteed to contain
+    // correct values. So `F::OFFSET` is still within the `F::Base` type.
+    let ptr: *const u8 = unsafe { ptr.add(F::OFFSET) };
+    let ptr: *const F::Type = ptr.cast::<F::Type>();
+    // SAFETY: The `Field` trait guarantees that at `F::OFFSET` we find a field of type `F::Type`.
+    unsafe { &*ptr }
+}
+```
+
+#### `field_of!` Macro
+
+Also added to `core::marker` is the following built-in macro:
+
+```rust
+pub macro field_of($Container:ty, $($fields:expr)+ $(,)?) {
+    /* built-in macro */
+}
+```
+
+It has the same syntax as the `offset_of!` macro and returns the respective [field type]. It emits an
+error in the following cases:
+
+```rust
+pub mod foo {
+    pub struct Foo {
+        bar: u32,
+        pub baz: i32,
+    }
+
+    // Error: unknown field `barr` of type `Foo`
+    type FooBar = field_of!(Foo, barr);
+}
+
+pub mod bar {
+    // Error: field `bar` of type `Foo` is private
+    type FooBar = field_of!(Foo, bar);
+}
+```
+
+#### `#[projecting]` Attribute
+
+The `#[projecting]` attribute can be put on a struct or union declaration. It requires that the
+type is `#[repr(transparent)]` and that there is a unique field that has non-zero size in general.
+It results in fields of the single non-zero sized type being considered fields of the outer type.
+
+So for example:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Container<T> {
+    inner: T,
+}
+
+struct Foo {
+    bar: i32,
+}
+```
+
+Now `Container<Foo>` has a [field type] associated with `bar` implementing `Field` with:
+- `Base = Container<Foo>`
+- `Type = Container<i32>`
+- `OFFSET = offset_of!(Foo, bar)`
+
+The same is true for the non-generic case:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Container {
+    inner: Foo,
+}
+
+struct Foo {
+    bar: i32,
+}
+```
```

Comment:
I don't understand, what is the `Field::Type` of `field_of!(Container, bar)` for the non-generic `Container`? There is no such thing as `Container<i32>` in this case.

Reply:
That is correct, I lost a bit of context for why I intended this to exist. But the natural choice for `Field::Type` would be `i32`:
```rust
impl Field for field_of!(Container, bar) {
    type Base = Container;
    type Type = i32;
    const OFFSET = offset_of!(Foo, bar);
}
```

But maybe this doesn't really make sense...

Reply:
with `OFFSET = offset_of!(Container, inner.bar)` this would make some sense but that's a pretty different usage than the impls on `UnsafeCell` and friends.

Reply:
Maybe it's just better to require a generic when using `#[projecting]` for the time being